tests/node_eviction_test.go

```diff
@@ -0,0 +1,238 @@
+package tests_test
+
+import (
+	"flag"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"kubevirt.io/kubevirt-ansible/tests"
+	ktests "kubevirt.io/kubevirt/tests"
+)
+
+var _ = Describe("Node Eviction", func() {
+
+	flag.Parse()
+
+	var (
+		simpleVmi   tests.VmManifest
+		vmBackedVmi tests.VmManifest
+		rsBackedVmi tests.VmManifest
+	)
+
+	ns := ktests.NamespaceTestDefault
+
+	BeforeEach(func() {
+		ktests.BeforeTestCleanup()
+
+		args := []string{"get", "node", "--selector=node-role.kubernetes.io/compute=true", "--output=jsonpath={.items..metadata.name}"}
+		output, err := ktests.RunOcCommand(args...)
+		Expect(err).ToNot(HaveOccurred())
+
+		var count int
+		nodes := strings.Fields(output)
+		for _, node := range nodes {
+			args := []string{"get", "node", node}
+			output, err := ktests.RunOcCommand(args...)
+			Expect(err).ToNot(HaveOccurred())
+
+			if !strings.Contains(output, "SchedulingDisabled") {
+				count += 1
+			}
+		}
+		if count < 2 {
+			Skip("not enough schedulable nodes for this testing")
+		}
+
+		simpleVmi = tests.VmManifest{
+			Name:     "vmi-ephemeral",
+			FilePath: "/tmp/vmi-ephemeral.yaml",
+			URL:      "https://raw.githubusercontent.com/kubevirt/kubevirt/master/cluster/examples/vmi-ephemeral.yaml",
+		}
+
+		vmBackedVmi = tests.VmManifest{
+			Name:     "vm-cirros",
+			FilePath: "/tmp/vm-cirros.yaml",
+			URL:      "https://raw.githubusercontent.com/kubevirt/kubevirt/master/cluster/examples/vm-cirros.yaml",
+		}
+
+		rsBackedVmi = tests.VmManifest{
+			Name:     "vmi-replicaset-cirros",
+			FilePath: "/tmp/vmi-replicaset-cirros.yaml",
+			URL:      "https://raw.githubusercontent.com/kubevirt/kubevirt/master/cluster/examples/vmi-replicaset-cirros.yaml",
+		}
+
+		tests.DownloadFromUrl(simpleVmi.URL, simpleVmi.FilePath)
+		tests.DownloadFromUrl(vmBackedVmi.URL, vmBackedVmi.FilePath)
+		tests.DownloadFromUrl(rsBackedVmi.URL, rsBackedVmi.FilePath)
+		tests.ReplaceStringInFile(simpleVmi.FilePath, tests.CirrosImageDev, tests.CirrosImageTest, -1)
+		tests.ReplaceStringInFile(vmBackedVmi.FilePath, tests.CirrosImageDev, tests.CirrosImageTest, -1)
+		tests.ReplaceStringInFile(rsBackedVmi.FilePath, tests.CirrosImageDev, tests.CirrosImageTest, -1)
+	})
+
+	Describe("VirtualMachineInstance Evictions", func() {
+		It("virtualmachine instance will not be re-scheduled", func() {
+			By("Create the virtualmachine instance")
+			args := []string{"create", "-n", ns, "-f", simpleVmi.FilePath}
+			_, err := ktests.RunOcCommand(args...)
+			Expect(err).ToNot(HaveOccurred())
+
+			By("Check that the virtualmachine instance is running")
+			args = []string{"get", "-n", ns, "vmi", simpleVmi.Name, "--template", "{{.status.phase}}"}
+			Eventually(func() (string, error) {
+				return ktests.RunOcCommand(args...)
```

Comment:
don't catch error here?

Reply:
no need, it asserts that error should be nil in the syntax.

Reply:
Hi @gouyang if you check the gomega lib here: https://github.com/shiywang/kubevirt-ansible/blob/a727b7cdbad9e32c7408598f158287dca5375d96/vendor/github.com/onsi/gomega/internal/asyncassertion/async_assertion.go#L84

the `pollActual` function only return first value, and in the "check-up" loop here:https://github.com/shiywang/kubevirt-ansible/blob/a727b7cdbad9e32c7408598f158287dca5375d96/vendor/github.com/onsi/gomega/internal/asyncassertion/async_assertion.go#L142, errors won't be throw out immediately, this might Okay for some cases, but for our case, we want to the errors being throw out asap instead of waitting for timeout.

```
•STEP: Wait until vmi ready
level=error timestamp=2018-08-16T02:34:24.617087Z pos=utils.go:1429 component=tests output="the server doesn't have a resource type \"abc\"\n" reason="exit status 1" msg="oc command failed: /usr/bin/oc get abc vm-cirros -o=jsonpath='{.status.phase}' -n kubevirt-test-default,"
level=error timestamp=2018-08-16T02:34:40.881898Z pos=utils.go:1429 component=tests output="the server doesn't have a resource type \"abc\"\n" reason="exit status 1" msg="oc command failed: /usr/bin/oc get abc vm-cirros -o=jsonpath='{.status.phase}' -n kubevirt-test-default,"
level=error timestamp=2018-08-16T02:34:57.191880Z pos=utils.go:1429 component=tests output="the server doesn't have a resource type \"abc\"\n" reason="exit status 1" msg="oc command failed: /usr/bin/oc get abc vm-cirros -o=jsonpath='{.status.phase}' -n kubevirt-test-default,"
level=error timestamp=2018-08-16T02:35:13.460318Z pos=utils.go:1429 component=tests output="the server doesn't have a resource type \"abc\"\n" reason="exit status 1" msg="oc command failed: /usr/bin/oc get abc vm-cirros -o=jsonpath='{.status.phase}' -n kubevirt-test-default,"
level=error timestamp=2018-08-16T02:35:29.719044Z pos=utils.go:1429 component=tests output="the server doesn't have a resource type \"abc\"\n" reason="exit status 1" msg="oc command failed: /usr/bin/oc get abc vm-cirros -o=jsonpath='{.status.phase}' -n kubevirt-test-default,"
level=error timestamp=2018-08-16T02:35:45.959108Z pos=utils.go:1429 component=tests output="the server doesn't have a resource type \"abc\"\n" reason="exit status 1" msg="oc command failed: /usr/bin/oc get abc vm-cirros -o=jsonpath='{.status.phase}' -n kubevirt-test-default,"
level=error timestamp=2018-08-16T02:36:02.193862Z pos=utils.go:1429 component=tests output="the server doesn't have a resource type \"abc\"\n" reason="exit status 1" msg="oc command failed: /usr/bin/oc get abc vm-cirros -o=jsonpath='{.status.phase}' -n kubevirt-test-default,"
level=error timestamp=2018-08-16T02:36:19.956530Z pos=utils.go:1429 component=tests output="the server doesn't have a resource type \"abc\"\n" reason="exit status 1" msg="oc command failed: /usr/bin/oc get abc vm-cirros -o=jsonpath='{.status.phase}' -n kubevirt-test-default,"

------------------------------
• Failure [131.600 seconds]
VM/VMI system test
/root/go/src/kubevirt.io/kubevirt-ansible/tests/expose_test.go:23
  Expose service via ssh and console VNC
  /root/go/src/kubevirt.io/kubevirt-ansible/tests/expose_test.go:31
    the VMI should be running [It]
    /root/go/src/kubevirt.io/kubevirt-ansible/vendor/github.com/onsi/ginkgo/ginkgo_dsl.go:365

    Timed out after 131.593s.
    Error: Unexpected non-nil/non-zero extra argument at index 1:
    	<*exec.ExitError>: &exec.ExitError{ProcessState:(*os.ProcessState)(0xc4202e64e0), Stderr:[]uint8(nil)}

    /root/go/src/kubevirt.io/kubevirt-ansible/tests/expose_test.go:51
------------------------------
•
------------------------------
• [SLOW TEST:44.839 seconds]
VM/VMI system test
```

Reply:
Why it need to throw out the error immediately? Since it's a loop, it's possible to get pass the error some time later. I think it's fine to just throw out the error after timeout.

for example, this is the result for a failed test, it looks okay to me.
```
    Timed out after 121.327s.
    Expected
        <string>: Failed
    to equal
        <string>: Running
```

Reply:
>Why it need to throw out the error immediately?

1. Timed out after 121.327s.
2. throw out error immediately.

because I think immediately < 121.327s, so we can let our code execute faster and CI pass faster.

Reply:
This is the behavior of `eventually` itself, if you think it's better to throw out the error immediately, can you start a discussion on gomega?